main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1694 @@
+///   2. We propagate variables.
+///   3. We perform trivial assignments where the left-hand variables does not occur in the RHS.
+///   4. We do full-blown Boolean unification with SVE.
+/// - We represent a conjunction with n >= 2 terms.
+///   - We group the terms into three: a set of constants, a set of variables, and a list of sub-terms.
+///   - We flatten conjunctions at least one level per call to `mkAnd`.
+/// - We normalize the representation of an equation t1 ~ t2.
+///   - We try to move single variables to the lhs.
+///   - We try to move ground terms (true/false/constants) to the RHS.
+///
+/// In the future, we could:
+/// - Explore change of basis.
+/// - Explore use slack variables (i.e. variables marked don't care, and use SAT).
+/// - Explore in detail how constraints are generated.
+///
+object FastBoolUnification {
+
+  /**
+    * The threshold for when a solution is considered too complex.
+    *
+    * The threshold is specified as an upper bound on the permitted connectives in a substitution.
+    *
+    * The threshold must be at least 600 for the test suite to pass.
+    *
+    * If a solution is too complex an [[TooComplexException]] exception is thrown.
+    */
+  private val Threshold: Int = 1_000
+
+  /**
+    * Enable debugging (prints information during Boolean unification).
+    */
+  private val Debugging: Boolean = true
+
+  /**
+    * Internal formatter. Used for debugging.
+    */
+  private val formatter: Formatter = Formatter.NoFormatter
+
+  /**
+    * Attempts to solve all the given unification equations `l`.
+    *
+    * Returns `Ok(s)` where `s` is a most-general unifier for all equations.
+    *
+    * Returns `Err(c, l, s)` where `c` is a conflict, `l` is a list of unsolved equations, and `s` is a partial substitution.
+    */
+  def solveAll(l: List[Equation]): Result[BoolSubstitution, (FastBoolUnificationException, List[Equation], BoolSubstitution)] = {
+    val solver = new Solver(l)
+    solver.solve()
+  }
+
+  /**
+    * A stateful phased solver for Boolean unification equations. The solver maintains two fields that change over time:
+    *
+    * - The current (pending) unification equations to solve.
+    * - The current (partial) substitution which represents the solution.
+    *
+    * @param l The input list of Boolean unification equations to solve.
+    */
+  private class Solver(l: List[Equation]) {
+
+    /**
+      * The current (pending) equations to solve.
+      *
+      * The list of pending equations decrease as the solver progresses.
+      *
+      * Note that the pending equations is not a strict subset of the original equations because they may be simplified during computation.
+      *
+      * If Boolean unification is successful, the list of pending equations will become empty at the end of the computation.
+      */
+    private var currentEqns: List[Equation] = l
+
+    /**
+      * The current substitution. Initially empty, but grows during computation.
```

Comment:
Maybe a note that it contains all variables of the input equations at the end?

Reply:
In what sense? If the equation is `x = x` then the empty substitution is a valid solution. 

(We don't have this particular optimization, but the point stands: the domain/codomain may not contain all variables.)